pkg/find-interface.go

```diff
@@ -0,0 +1,105 @@
+package impl
+
+import (
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"strconv"
+	"strings"
+
+	"golang.org/x/tools/imports"
+)
+
+func findTopTypeDecl(id string, f *ast.File) (*ast.GenDecl, *ast.TypeSpec) {
+	for _, decl := range f.Decls {
+		decl, ok := decl.(*ast.GenDecl)
+		if !ok || decl.Tok != token.TYPE {
+			continue
+		}
+		for _, spec := range decl.Specs {
+			switch spec := spec.(type) {
+			case *ast.TypeSpec:
```

Comment:
switches with a single case can be simplified with a negated if that exits/continues early. That way, the rest of the code will be less indented.

This applies in a few other cases throughout the PR.

Reply:
I was on the fence about this myself, IIRC. It's been a while of course, but I think I was trying to maintain consistency with the rest of the code since there are several similar cases where multiple types are expected.

In general, I find the type switch syntax more readable than the `!ok` early exit, and for future additions (e.g. I'm not sure how this exact code will interact with type aliases. I would assume it needs to / will need to understand them at some point, but they would not be a `*ast.TypeSpec`. For maintenance, the type switch makes it much less work to refactor.

Reply:
Oh, if this was for consistency with existing code, then sure. Not familiar with the codebase :)

Reply:
I think it was more self-consistency, since the codebase didn't have any AST handling previously. There are a few cases (https://github.com/josharian/impl/pull/16/files#diff-e2677c2842f04650fa076924ce87c37eR98) where we handle several possible types in the AST code.

Not sure if that changes your opinion, but if so I'm always open to being convinced to switch things up to keep it more maintainable. 😄